File: Towers Reaching Both Stations.cpp
//Problem: Towers Reaching Both Stations
//geeksforgeeks problem of the day -> 8th July 2026
//C++ CODE
class Solution {
  public:
    vector<int>dis={0,1,0,-1,0};
    void dfs(int i, int j, vector<vector<int>>& mat,
             vector<vector<bool>>& vis) {

        vis[i][j] = true;
        int n=mat.size();
        int m=mat[0].size();

        for (int d=0;d<4;d++) {
            int ni = i + dis[d];
            int nj = j + dis[d+1];

            if (ni < 0 || nj < 0 || ni >= n || nj >= m)
            {
                continue;
            }

            if (vis[ni][nj])
            {
                continue;
            }
            
            if (mat[ni][nj] >= mat[i][j])
            {
                dfs(ni, nj, mat, vis);
            }
        }
    }
    
    int countCoordinates(vector<vector<int>>& mat) {
        // code here
        int n = mat.size();
        int m = mat[0].size();

        vector<vector<bool>> stationP(n, vector<bool>(m, false));
        vector<vector<bool>> stationQ(n, vector<bool>(m, false));

        for (int j = 0; j < m; j++)
        {
            if (!stationP[0][j])
            {
                dfs(0, j, mat, stationP);
            }
        }

        for (int i = 0; i < n; i++)
        {
            if (!stationP[i][0])
            {
                dfs(i, 0, mat, stationP);
            }
        }

        for (int j = 0; j < m; j++)
        {
            if (!stationQ[n - 1][j])
            {
                dfs(n - 1, j, mat, stationQ);
            }
        }

        for (int i = 0; i < n; i++)
        {
            if (!stationQ[i][m - 1])
            {
                dfs(i, m - 1, mat, stationQ);
            }
        }

        int ans = 0;

        for (int i = 0; i < n; i++) 
        {
            for (int j = 0; j < m; j++) 
            {
                if (stationP[i][j] && stationQ[i][j])
                {
                    ans++;
                }
            }
        }

        return ans;
    }
};

File: Towers Reaching Both Stations.py
#Problem: Towers Reaching Both Stations
#geeksforgeeks problem of the day -> 8th July 2026
#Python Code
from collections import deque

class Solution:
    def countCoordinates(self, mat):
        n = len(mat)
        m = len(mat[0])

        def bfs(starts):
            vis = [[False] * m for _ in range(n)]
            q = deque()

            for r, c in starts:
                if not vis[r][c]:
                    vis[r][c] = True
                    q.append((r, c))

            directions = [(-1, 0), (1, 0), (0, -1), (0, 1)]

            while q:
                r, c = q.popleft()

                for dr, dc in directions:
                    nr, nc = r + dr, c + dc

                    if (0 <= nr < n and 0 <= nc < m and
                        not vis[nr][nc] and
                        mat[nr][nc] >= mat[r][c]):

                        vis[nr][nc] = True
                        q.append((nr, nc))

            return vis

        # Station P (top row + left column)
        p_starts = [(0, j) for j in range(m)] + [(i, 0) for i in range(n)]

        # Station Q (bottom row + right column)
        q_starts = [(n - 1, j) for j in range(m)] + [(i, m - 1) for i in range(n)]

        p = bfs(p_starts)
        q = bfs(q_starts)

        ans = 0
        for i in range(n):
            for j in range(m):
                if p[i][j] and q[i][j]:
                    ans += 1

        return ans